llama_index/callbacks/open_inference_callback.py

```diff
@@ -0,0 +1,244 @@
+"""
+Callback handler for storing generation data in OpenInference format.
+OpenInference is an open standard for capturing and storing AI model inferences.
+It enables production LLMapp servers to seamlessly integrate with LLM
+observability solutions such as Arize and Phoenix.
+
+For more information on the specification, see
+https://github.com/Arize-ai/open-inference-spec
+"""
+
+import importlib
+import uuid
+from dataclasses import dataclass, field, fields
+from datetime import datetime
+from types import ModuleType
+from typing import TYPE_CHECKING, Any, Callable, Dict, Iterable, List, Optional, TypeVar
+
+from llama_index.callbacks.base import BaseCallbackHandler
+from llama_index.callbacks.schema import CBEventType, EventPayload
+
+if TYPE_CHECKING:
+    from pandas import DataFrame
+
+
+OPENINFERENCE_COLUMN_NAME = "openinference_column_name"
+Embedding = List[float]
+
+
+def _generate_random_id() -> str:
+    """Generates a random ID.
+
+    Returns:
+        str: A random ID.
+    """
+    return str(uuid.uuid4())
+
+
+@dataclass
+class QueryData:
+    """
+    Query data with column names following the OpenInference specification.
+    """
+
+    id: str = field(
+        default_factory=_generate_random_id,
+        metadata={OPENINFERENCE_COLUMN_NAME: ":id.str:"},
+    )
+    timestamp: Optional[str] = field(
+        default=None, metadata={OPENINFERENCE_COLUMN_NAME: ":timestamp.iso_8601:"}
+    )
+    query_text: Optional[str] = field(
+        default=None,
+        metadata={OPENINFERENCE_COLUMN_NAME: ":feature.text:prompt"},
+    )
+    query_embedding: Optional[Embedding] = field(
+        default=None,
+        metadata={OPENINFERENCE_COLUMN_NAME: ":feature.[float].embedding:prompt"},
+    )
+    response_text: Optional[str] = field(
+        default=None, metadata={OPENINFERENCE_COLUMN_NAME: ":prediction.text:response"}
+    )
+    document_ids: List[str] = field(
+        default_factory=list,
+        metadata={
+            OPENINFERENCE_COLUMN_NAME: ":feature.[str].retrieved_document_ids:prompt"
```

Comment:
It's up to you, but I find the document_ids name here a little misleading -- it's actually node ids

Reply:
Dang, is there a way to get the document IDs themselves? I am seeing that in the actually JSON index.

Reply:
I'm noticing that the retrieved node IDs are not present in the original index 🧐

Reply:
We are trying to using the document ID as a join key on the original system of record.

Reply:
Yea, input documents are broken into nodes, all with new ids

Each node has a little copy lf the original document id under `node.ref_doc_id`

Reply:
Many nodes from the same document could be retrieved

Reply:
@axiomofjoy just making sure you saw these two responses 👌

I saw you added some hash stuff. I'm not sure what the intended use case for logging the ids is, but the node ID is the real id that is stored in the docstore.

The original document is lost when it's ingested into the index, as it is broken down into nodes for retrieval. But you can trace which document a node belonged to by using node.ref_doc_id

Reply:
Thanks @logan-markewich! I'll take a second look today and will follow up with additional questions.

Reply:
Renamed "document" to "node" in ef6c4ed13dd54a8fb96920ff08fb569c9112d45e. Thanks for the callout 🙏

Reply:
I'm finding that when I load a pre-built index from storage, the `node.node_id` field does not match any of the IDs in the index. Is it expected behavior?